tests/gtest/avifallocationtest.cc

```diff
@@ -0,0 +1,169 @@
+                 std::numeric_limits<typeof(avifImage::height)>::max(), 12,
+                 AVIF_RESULT_INVALID_ARGUMENT);
+  TestAllocation(1, 1, std::numeric_limits<typeof(avifImage::depth)>::max(),
+                 AVIF_RESULT_UNSUPPORTED_DEPTH);
+}
+
+void TestEncoding(uint32_t width, uint32_t height, uint32_t depth,
+                  avifResult expected_result) {
+  testutil::AvifImagePtr image(avifImageCreateEmpty(), avifImageDestroy);
+  ASSERT_NE(image, nullptr);
+  image->width = width;
+  image->height = height;
+  image->depth = depth;
+  image->yuvFormat = AVIF_PIXEL_FORMAT_YUV444;
+
+  // This is a fairly high number of bytes that can safely be allocated in this
+  // test. The goal is to have something to give to libavif but libavif should
+  // return an error before attempting to read all of it, so it does not matter
+  // if there are fewer bytes than the provided image dimensions.
+  static constexpr uint64_t kMaxAlloc = 2147483647;
+  size_t num_allocated_bytes;
+  if ((uint64_t)image->width * image->height >
+      kMaxAlloc / (avifImageUsesU16(image.get()) ? 2 : 1)) {
+    num_allocated_bytes = kMaxAlloc;
+  } else {
+    num_allocated_bytes =
+        image->width * image->height * (avifImageUsesU16(image.get()) ? 2 : 1);
+  }
+
+  // Initialize pixels as 16b values to make sure values are valid for 10
+  // and 12-bit depths. The array will be cast to uint8_t for 8-bit depth.
+  std::vector<uint16_t> pixels(num_allocated_bytes / sizeof(uint16_t), 400);
+  uint8_t* bytes = reinterpret_cast<uint8_t*>(pixels.data());
+  // Avoid avifImageAllocatePlanes() to exercise the checks at encoding.
+  image->imageOwnsYUVPlanes = AVIF_FALSE;
+  image->imageOwnsAlphaPlane = AVIF_FALSE;
+  image->yuvRowBytes[AVIF_CHAN_Y] = image->width;
+  image->yuvPlanes[AVIF_CHAN_Y] = bytes;
+  image->yuvRowBytes[AVIF_CHAN_U] = image->width;
+  image->yuvPlanes[AVIF_CHAN_U] = bytes;
+  image->yuvRowBytes[AVIF_CHAN_V] = image->width;
+  image->yuvPlanes[AVIF_CHAN_V] = bytes;
+  image->alphaRowBytes = image->width;
+  image->alphaPlane = bytes;
+
+  // Try to encode.
+  testutil::AvifEncoderPtr encoder(avifEncoderCreate(), avifEncoderDestroy);
+  ASSERT_NE(encoder, nullptr);
+  encoder->speed = AVIF_SPEED_FASTEST;
+  testutil::AvifRwData encoded_avif;
+  ASSERT_EQ(avifEncoderWrite(encoder.get(), image.get(), &encoded_avif),
+            expected_result);
+}
+
+TEST(EncodingTest, MinimumValid) { TestAllocation(1, 1, 8, AVIF_RESULT_OK); }
+
+TEST(EncodingTest, MaximumValid) {
+  TestEncoding(65535, 1, 12, AVIF_RESULT_OK);
+  TestEncoding(1, 65535, 12, AVIF_RESULT_OK);
+  // TestEncoding(16384, 8096, 12, AVIF_RESULT_OK);  // Too slow.
+}
+
+TEST(EncodingTest, MinimumInvalid) {
+  TestEncoding(0, 1, 8, AVIF_RESULT_NO_CONTENT);
+  TestEncoding(1, 0, 8, AVIF_RESULT_NO_CONTENT);
+  TestEncoding(1, 1, 0, AVIF_RESULT_UNSUPPORTED_DEPTH);
+  TestEncoding(65536, 1, 12, AVIF_RESULT_ENCODE_COLOR_FAILED);
```

Comment:
and AVIF_DEFAULT_IMAGE_SIZE_LIMIT + 1 here?

Reply:
Changed to maximum AV1 dimension + 1.